apps/labs/posts/libsf-error-state.md

```diff
@@ -0,0 +1,744 @@
+// that consumers know the symbol is defined inside the DLL. In all other cases,
+// the macro expands to nothing.
+// Note: SCIPY_DLL_{EX,IM}PORTS are set in scipy/special/meson.build
+#if defined(SCIPY_DLL_EXPORTS)
+    #define SCIPY_DLL __declspec(dllexport)
+#elif defined(SCIPY_DLL_IMPORTS)
+    #define SCIPY_DLL __declspec(dllimport)
+#else
+    #define SCIPY_DLL
+#endif
+```
+
+As soon as I had a chance, I fired up a Windows VM again and put together [a PR](https://github.com/scipy/scipy/pull/20937)
+implementing Axel's solution. After a couple misteps, MSVC builds were working again. There would be no need to push
+back the release date. A couple weeks later, fellow Quansight Labs member and LFortran/LPython core developer Gagandeep
+Singh (czgdp1807) submitted [a PR](https://github.com/scipy/scipy/pull/20985) to add an MSVC CI job, plugging the gap in
+SciPy's coverage.
+
+### Thread safety
+
+In the introduction I'd mentioned that CPython's global interpreter lock (GIL) makes it easier to extend Python with C
+or other compiled languages since one doesn't need to worry about the thread safety of wrapped code. Still, being able
+to have only one thread in a running Python process execute Python code at a time is greatly limiting for multicore
+applications. In October 2023, a Python Enhancement Proposal (PEP) was accepted to make the GIL optional. This proposal,
+[PEP 703](https://peps.python.org/pep-0703/), is well worth reading for its in-depth summary of the surrounding
+issues. CPython 3.13 launched with an optional
+[free-threaded mode](https://docs.python.org/3.13/whatsnew/3.13.html#free-threaded-cpython)
+which supports running with the GIL disabled. Free-threading in Python offers great promise, but to take advantage of
+it, extensions need to be rewritten with thread safety in mind.
+
+`libsf_error_state` is obviously not thread safe. There is a global array carrying the current state of special
+function error handling policies with nothing to stop competing threads from trying to access the same memory
+location. Simultaneous modifications could even leave an entry in some corrupt and indeterminate state. This situation
+is known as a data race and leads to undefined behavior in C and C++ []. Weird things can happen when there is a data race.
+
+The latest tale in the saga of `libsf_error_state` is [a PR](https://github.com/scipy/scipy/pull/21956) from Edgar
+Margffoy (andfoy)—a member of Quansight Labs' free-threaded Python team—to ensure thread safety by declaring the array
+`sf_error_actions` thread local. This eliminates the data-race by making it so each thread gets its own separate copy of
+the array. Edgar and the others on the free-threaded Python team have been doing great work improving support for
+free-threaded Python across the ecosystem for much of this past year.
+
+In a curious reversal, it is now (the `win32` flavor of) MinGW that is causing trouble due to lack of proper threading
+support [].
```

Comment:
empty brackets. I think the sentence should be continued with something like "since MinGW does not support thread local variables".

Reply:
I was basing it on this from Edgar

*It seems that using std::mutex also causes the same MinGW threading issues to surface.*

and part of your reply

* MinGW-w64 comes in two build flavors: posix and win32. The posix builds have threading support, the win32 ones don't*

It seemed like something deeper might be going on, since even the mutex isn't working. I don't really understand the underlying issue though.

Reply:
I don't think there's a need to explain in detail, just the basic reason (MinGW doesn't have support).